pkg/spanconfig/spanconfigsqlwatcher/sql_watcher.go

```diff
@@ -0,0 +1,453 @@
+					// config entry only needs to be removed when the descriptor is
+					// deleted, not when it is dropped, so we set the Deleted flag only if
+					// no descriptor with the given ID exists.
+					_, err := descsCol.GetImmutableDescriptorByID(ctx, txn, descID, tree.CommonLookupFlags{
+						AvoidCached:    true,
+						IncludeDropped: true,
+					})
+					if err != nil {
+						if errors.Is(err, catalog.ErrDescriptorNotFound) {
+							update.Deleted = true
+						} else {
+							return err
+						}
+					}
+				}
+				updates = append(updates, update)
+			}
+			return nil
+		}); err != nil {
+		return err
+	}
+
+	// Send all the entries on the updatesCh.
+	for _, update := range updates {
+		select {
+		case <-s.stopper.ShouldQuiesce():
+		case <-ctx.Done():
+		case updatesCh <- update:
+		}
+	}
+	return nil
+}
+
+// generateSpanConfigurations generates the span configurations corresponding to
+// the provided list of IDs. It uses a transactional view of system.zones and
+// system.descriptors to do so.
+func (s *SQLWatcher) generateSpanConfigurations(
+	ctx context.Context, txn *kv.Txn, ids descpb.IDs,
+) ([]roachpb.SpanConfigEntry, error) {
+	copyKey := func(k roachpb.Key) roachpb.Key {
+		k2 := make([]byte, len(k))
+		copy(k2, k)
+		return k2
+	}
+
+	ret := make([]roachpb.SpanConfigEntry, 0, len(ids))
+	for _, id := range ids {
+		zone, err := sql.GetHydratedZoneConfigForTable(ctx, txn, s.codec, id)
+		if err != nil {
+			return nil, err
+		}
+		spanConfig, err := zone.ToSpanConfig()
+		if err != nil {
+			return nil, err
+		}
+
+		tablePrefix := s.codec.TablePrefix(uint32(id))
+		prev := tablePrefix
+		for i := range zone.SubzoneSpans {
+			// We need to prepend the tablePrefix to the spans stored inside the
+			// SubzoneSpans field because we store the stripped version here for
+			// historical reasons.
+			span := roachpb.Span{
+				Key:    append(tablePrefix, zone.SubzoneSpans[i].Key...),
+				EndKey: append(tablePrefix, zone.SubzoneSpans[i].EndKey...),
+			}
+
+			{
+				// The zone config code sets the EndKey to be nil before storing the
+				// proto if it is equal to `Key.PrefixEnd()`, so we bring it back if we
+				// required.
+				if zone.SubzoneSpans[i].EndKey == nil {
+					span.EndKey = span.Key.PrefixEnd()
+				}
+			}
+
+			// If there is a "hole" in the spans covered by the subzones array we fill
+			// it using the parent zone configuration.
+			if !prev.Equal(span.Key) {
+				ret = append(ret,
+					roachpb.SpanConfigEntry{
+						Span:   roachpb.Span{Key: copyKey(prev), EndKey: copyKey(span.Key)},
+						Config: spanConfig,
+					},
+				)
+			}
+
+			// Add an entry for the subzone.
+			subzoneSpanConfig, err := zone.Subzones[zone.SubzoneSpans[i].SubzoneIndex].Config.ToSpanConfig()
+			if err != nil {
+				return nil, err
+			}
+			ret = append(ret,
+				roachpb.SpanConfigEntry{
+					Span:   roachpb.Span{Key: copyKey(span.Key), EndKey: copyKey(span.EndKey)},
```

Comment:
Why not simply use `span`, instead of desconstructing + reconstructing a roachpb.Span.